plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroIdentityPlugin.java

```diff
@@ -101,6 +109,36 @@ public TokenManager getTokenManager() {
     }
 
     @Override
+    public UnaryOperator<RestHandler> getRestHandlerWrapper(ThreadContext threadContext) {
+        return AuthcRestHandler::new;
+    }
+
+    class AuthcRestHandler extends DelegatingRestHandler {
+        public AuthcRestHandler(RestHandler original) {
+            super(original);
+        }
+
+        @Override
+        public void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
```

Comment:
This looks like it is the old restHandler logic. I take it the idea is that each plugin would implement their own equivalent to take care of things now?

Reply:
Yes exactly, the IdentityPlugin's role is to figure out how to provide information about the current subject. For identity-shiro plugin (and the security plugin) they perform auth using the `ActionPlugin.getRestHandlerWrapper` extension point and hydrate the ThreadContext with information about the authenticated user. This uses the same auth from the ShiroPlugin, but moves where it is taking place. Its still within the RestController.

Reply:
I see. Can you clarify what the benefit of using the unaryOperator is then versus letting OpenSearch decide when to auth?

Reply:
See the comment chain starting from [here](https://github.com/opensearch-project/OpenSearch/pull/15430#discussion_r1736957906). The problem with the existing code behind the feature flag in the RestController is the rigidity. 

The benefit of `ActionPlugin.getRestHandlerWrapper()` is the flexibility of having access to the RestRequest. There are many ways that auth info can be shipped on a RestRequest, not only the AUTHORIZATION http header.

For instance there is also:

- url param with jwt auth
- client cert auth
- Customization of the HTTP header that auth info could be sent on
- Proxy auth which uses XFF headers

With the current implementation of RestTokenExtractor, it assumes that all auth info is sent on the Authorization header. That may be true for shiro-identity plugin, but is not the case for other identity plugins.

Reply:
Marking thread as resolved.